### DISCUSSION TITLE
Create chart with unkown columns

### DISCUSSION THREAD

**Discussion by that-one-tom** [Budibase/budibase#8442] [Help] [answered]
Hi all, the [documentation ](https://docs.budibase.com/docs/chart#many-data-columns) mentions

> If you do not know ahead of time how many labels or data columns you will be displaying, then you will need to use JavaScript to loop over your API data.

I am working on a screen where users can pick which data series they want to compare in a line chart. The transformation in the data provider can handle this just fine (and returns one label column + 1...n data columns with changing names), but I can't get the line chart to work with this setup.

Did I perhaps overlook something or is this simply not supported at this time?

Thank you (and also thanks for Budibase, it's quite fun to work with)!

**Comment by that-one-tom** [comment]
I've now created a quick and dirty plugin to let me create such a chart: https://github.com/that-one-tom/budibase-chartjs

Still curious to understand whether I might have just missed something here. Thanks everyone!

**Comment by melohagan** [comment]
Hey @that-one-tom 

Really glad you were able to create a custom component to meet your needs!

If you provide an example of your API response, and what you want the chart to display, then I can have a go at getting that to work with the line chart.

**Reply by that-one-tom** [reply]
Hi @melohagan, thanks so much for looking into this! So I am using a PostgreSQL. My database table holding a bunch of pricing information captured at different timestamps looks roughly like this:

```
 id | timestamp                | product_id | price
----+--------------------------+------------+-------
  1 | 2022-11-09T08:00:00.000Z |          1 |  9.99
  2 | 2022-11-09T08:00:00.000Z |          2 | 11.99
  3 | 2022-11-09T08:00:00.000Z |          3 | 10.99
  4 | 2022-11-10T08:00:00.000Z |          1 |  9.99
  5 | 2022-11-10T08:00:00.000Z |          2 |  7.99
  6 | 2022-11-10T08:00:00.000Z |          3 | 10.99
  7 | 2022-11-11T08:00:00.000Z |          1 |  9.99
  8 | 2022-11-11T08:00:00.000Z |          2 |  7.99
  9 | 2022-11-11T08:00:00.000Z |          3 | 12.99
```

Now buyers (or whoever else needs to compare prices) can select which products they wan't to keep an eye on. I would then use a query (plus possibly a transformation) to select the prices based on the product ID, add a join to fetch the product names and in the end return a result like this:

```
 Date       | Product A | Product B
------------+-----------+-----------
 2022-11-09 |      9.99 |     11.99
 2022-11-10 |      9.99 |      7.99
 2022-11-11 |      9.99 |      7.99
```

The possible product names and IDs are also unknown at build time (as new products might get added to the database over time). Such a table can still be displayed perfectly fine in Budibase, even if I don't know which or how many products have been selected by the user. 

I could, however, not get a line chart like this to work in Budibase. It seemed to me the values in the `Data Cols.` field of the Line Chart have to be set at build time (so I couldn't set something like "all columns except ...", or provide the list of columns through Handlebars or JavaScript).

![grafik](https://user-images.githubusercontent.com/19203795/201342800-97062b59-6682-48d3-8fad-ce6f2da93520.png)

There is a very good chance I have simply missed something, so any help or suggestions would be much appreciated

**Reply by melohagan** [reply]
Hey Tom, having had a look at this I think you're right! It doesn't seem that there is a way to handle an unknown number of data columns at the moment.

**Reply by melohagan** [reply]
Raised a feature request here: https://github.com/Budibase/budibase/issues/8652

**Reply by that-one-tom** [reply]
Thanks so much for confirming and creating the feature request!

**Reply by mmays77** [reply, marked as answer]
I know this is a very late response to your question, but what you are looking for is something I researched and solved for myself as described below. Apologies if this is an inelegant or outdated way to solve this, but it is working for me. If there is a better way, please let me know.

The need is a pivot table, and for budibase to handle it properly, you need rows of dates and columns with your data values in them. There are a few external tools that will do these kinds of pivot tables, including spreadsheets and data tools, and you can navigate to them using a button, but this is not ideal. 

To stay within the budibase environment, you can create a simple rest server in python or another language that generates dynamic sql to create the pivot table that you need. If you issue a "get" command to a python rest server that uses flask or django, python can do selects against the data tables to determine the rows and columns and create the sql that you want. The python then submits the generated dynamic SQL to the database engine and returns the result to budibase as a rest query in response to the "get" request. 

The program will look at each row in a table, and add the value to a column named "Product A" or "Product B". The first column is your date and is used with a group by at the end of the dynamic SQL.

The following SQL does something similar summarizing by TransactionDate. This code can be embedded in a python rest server and executed dynamically, so the rows based on dates are always current.


SELECT
  GROUP_CONCAT(DISTINCT
    CONCAT(
      'SUM(CASE WHEN Category = ''', Category, ''' AND SubCategory = ''', SubCategory,
      ''' THEN Amount ELSE 0 END) AS ', 
      CONCAT('total_', REPLACE(CONCAT(AccountName, '_', Category, '_', SubCategory), ' ', '_'))
    )
  ) INTO @sql
FROM YourTableName;

SET @sql = CONCAT('SELECT TransactionDate, ', @sql, ' 
                   FROM YourTableName 
                   GROUP BY TransactionDate');

PREPARE stmt FROM @sql;
EXECUTE stmt;
DEALLOCATE PREPARE stmt;


In a python rest server running this as embedded code, you do not need to deallocate the statement because it is done automatically.

I hope this helps. The only problem I am having is it seems that budibase line chart is discarding trailing zeros from the returned rest query values, so currency looks funky when 48.10 gets truncated to 48.1. I have not figure out how to fix this yet.
